squape/object_tree.py

```diff
@@ -0,0 +1,215 @@
+import math
+import types
+
+try:
+    import squish
+except ImportError:
+    import squishtest as squish
+
+import object
+
+
+def children(object_name: any, selector: dict = None) -> tuple:
+    """Filter the direct children of the specified object by the specified selector
+
+    Args:
+        object_name (any): symbolic name, real name, or object reference to the parent
+        selector (dict, optional): The selector is a dictionary of properties,
+        that must match for objects to be included into the result. 
+        The values of the selector can be lambda functions as well.
+        In that case, the verification would consist of passing the candidate's key attribute into that function. 
+        The return value should always be bool.
+        Defaults to {}, which means all objects pass the verification.
+
+    Returns:
+        tuple: children objects that met the selector criteria
+
+    Examples:
+        >>> children(object_name, {'type' : Button, 'visible' : True})
+    """
+    if selector is None:
+        selector = {}
+    object_reference = _get_object_reference(object_name)
+    children = object.children(object_reference)
+    return tuple(filter(lambda x: _is_matching(x, selector), children))
+
+
+def find(object_name: any, selector: dict = None, max_depth: int = None) -> tuple:
+    """Recursively filter all the children of the specified object
+    by the specified selector
+
+    Args:
+        object_name (any): symbolic name, real name, or object reference to the parent
+        selector (dict, optional): The selector is a dictionary of properties,
+        that must match for objects to be included into the result. 
+        The values of the selector can be lambda functions as well.
+        In that case, the verification would consist of passing the candidate's key attribute into that function. 
+        The return value should always be bool.
+        Defaults to {}, which means all objects pass the verification.
+        max_depth (int): defines maximum depth in the object structure that should be
+        while looking for children.
+        Defaults to None, which mean there is no depth limit.
+
+    Returns:
+        tuple: the result of search among the object tree.
+
+    Examples:
+        ```python
+        find(object)
+        find(object, {'type' : 'ToolBar'})
+        find(object, max_depth=5)
+        find(object, {'visible' : True}, max_depth=3)
+        ```
+    """
+    if max_depth is None:
+        max_depth = math.inf
+    if max_depth <= 0:
+        return ()
+    if selector is None:
+        selector = {}
+
+    object_reference = _get_object_reference(object_name)
+    children = ()
+
+    for child in object.children(object_reference):
+        if _is_matching(child, selector):
+            children += (child,)
+        children += find(child, selector, max_depth - 1)
+
+    return children
+
+
+def find_ancestor(object_name: any, selector: dict = None):
```

Comment:
@pawtom shall we make the selector a  required parameter? Without selector shit function behaves like `parent()`

Reply:
@EduardKaverinskyi please make the selector parameter required